src/action.rs

```diff
@@ -1,7 +1,48 @@
-//! The different actions that can be done.
+//! The different actions that can be executed via any given key.
 
 use crate::key_code::KeyCode;
 
+/// The different types of actions we support for key macros
+#[non_exhaustive] // Definitely NOT exhaustive!  Let's add more! Mouse events maybe? :)
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum SequenceEvent {
+    /// A keypress/keydown
+    Press(KeyCode),
+    /// Key release/keyup
+    Release(KeyCode),
+    /// A shortcut for `Press(KeyCode), Release(KeyCode)`
+    Tap(KeyCode),
+    /// For sequences that need to wait a bit before continuing
+    Delay {
+        /// A delay (in ticks) to wait before executing the next SequenceEvent
+        since: u32,
+        /// Number of ticks to wait before removing the Delay
+        ticks: u32,
+    },
+    /// A marker that indicates there's more of the macro than would fit
+    /// in the 'sequenced' ArrayDeque
+    Continue {
+        /// The current chunk
+        index: usize,
+        /// The full list of Sequence Events (that aren't Continue())
+        events: &'static [SequenceEvent],
+    },
+    /// Cancels the running sequence and can be used to mark the end of a sequence
+    /// instead of using a number of Release() events
+    Complete,
+}
+
+impl SequenceEvent {
+    /// Returns the keycode associated with the given Press/Release event
+    pub fn keycode(&self) -> Option<KeyCode> {
```

Comment:
That's never used?

Reply:
Well, I used it in *debugging* heh heh.  I can get rid of it if you want.  It's not important.  I'm used to Python where metaprogramming and introspection are easy so in Rust I seem to have developed this habit of adding methods of introspection to everything so I can get a view into what's going on whenever I need it.

Reply:
derive(Debug) is your friend